### ISSUE TITLE
Ugly rendering of addr:flats

### ISSUE THREAD

**Issue by sanderd17** [gravitystorm/openstreetmap-carto#4160]
### Expected behavior

The `addr:flats` tag is used to encode apartment numbers into OSM. Big apartments will usually have a big number of flats, and thus a long list under the `addr:flats` tag.

In contrast to `addr:unit` (which is meant to address a single unit), `addr:flats` is usually a list of apartments. These apartments are usually also on top of each other.

It's geographical info that can be used for address completion/verification, but IMO is very unsuited for rendering on a 2D map.

### Actual behavior

Ugly long lists of numbers which makes it almost impossible to find the housenumbers.

### Links and screenshots illustrating the problem

![afbeelding](https://user-images.githubusercontent.com/562171/84490029-debf9b80-aca2-11ea-94b4-b31cb6e78e86.png)

![afbeelding](https://user-images.githubusercontent.com/562171/84490424-71f8d100-aca3-11ea-870e-f197eab13133.png)

### Solution

Revert #3988 and display only `addr:unit` for separate addressed building units.

**Comment by Adamant36** [comment]
The Wiki says (and I'm pretty sure that the discuss said to) that it's suppose to be used for a range of numbers. So, your example is probably miss-tagging. Doing a quick glance at TagInfo it seems like values aren't that way. You should take it as an opportunity improve tagging, like would be done in any other similar situation. It's not like the name, ref, or any other similar tags don't have the same problem to some degree.

**Comment by sanderd17** [comment]
Here are the links to the shown screenshot locations. Please tell me which ones aren't correctly mapped according to you.

If anything, the default rendering here seems to suggest we need to delete this valid data, or invent yet another tag (`addr:flat-list`, `addr:subhousenumbers`, ...)

https://www.openstreetmap.org/#map=19/50.94107/3.06177
https://www.openstreetmap.org/#map=19/50.87525/4.69097

**Comment by Adamant36** [comment]
Personally, both links look like junk tagging to me. At least as far as the definition of the tag goes. Its easy to find a edge cases to cherry pick to support whatever perspective, you can do it for every tag. But like I said 99% of the values are used correctly. So its not worth reverting IMO. As far as the suggestion of a flat list tag, this isnt the place for tagging discussions. Maybe message the person that made used the tag wrong though to see if they will fix it or tell you how though. That's what id do. A huge part of OSM is improving tagging in situations where we can. So, dont treat me like its a bad suggestion ;)

**Comment by sanderd17** [comment]
I was one of those taggers, tagged long before it was rendered. But I veryfied my tagging with someone elses tagging, and it matches what I did.

According to the wiki, our tagging is conform, the wiki even has such a long list as example:

>If there is gap in numeration, mark numbers and intervals with semicolon: addr:flats=3-7;10;14;16-18.

I don't know where you get the number of 99% good tagging. If I look around, I see at least 60% long lists.

**Comment by sanderd17** [comment]
I extracted a list of all values used in the city of Ghent (a region where I don't map, and which I didn't check before).

Most of the short examples are wrong tagging: either it lists the number of flats, which should be under `building:flats`, or it lists a single flat number, which should be under `addr:unit`. 

If you filter those single values out, the vast majority seems rather inconvenient to render.

Note: searched via overpass, and processed the JSON to get only the values out: http://overpass-turbo.eu/s/V3y

<details>
```
1;101
1-14
1;101
1
1-8
1-8
1
1
1
1-6;10;20;30;32;34;42-45;48-50;53-54;57-58;60-62;70-77;42A-42F
101-102;201-206;301-308
1-2;11-19;21-29;31-39
1-2;11-12
1-2;11-12;21-22
1-2;11-12;21-22
11;21
1-4;101
1;101;201;301
1;101;201
1;101-102;201-202
1;101;201;301
1;101;201
1-2
1-2;101-103;201-203;301-303
1;101
1;101;201
1;101;201
1;101
1-2;101-102;201-202
1-2;101;201;301
1;101;201
1;201;301
1-8
1-8
1-8
1-8
1;101;102
1;101-103
1;101-103;201-203;301-302
1-2;101-102;201-202
1;101
1-3;101-102;201-202;301-304;401-406
1;101-102;201
1-2;101-102;201-202;301-302;401-402;501-503;601-602;701
1-2;101-102;201-202;301-302;401-402;501-503;601-602;701
1-2;101-102;201-202;301-302;401-402;501-503;601-602;701
1;101;201-202
101;201
101-104;201-204
1;101;301
1;101;201
1-6;101-107;201-207;301-307
1;101-102
1;101;201;301
1;101-102;201-202;301-302
1;101-102;201-202;301-302;401
1;101-103;201-203;301-303
1;101;201
1-2;101;201
201
1;101
1;101;201
1-2;101-102
1-3;101-103;201-203
1;101;201
1;101;201
101
1;101;201
101-104;201-204
101;201
1;101-102;201-202;301
201-202
101-102;201-202;301
2
1-3;101-104;201-204;301-304
101
1;101-102;201-202;301-302
1;101;201;301
1-3;101-104;201-204;301-304
1-3;101-104;201-204;301-304
1-3;101-104;201-204;301-304
1-3;101-104;201-204;301-304
101-103;201-203;301-303;401-403
1;101;201
1-3;101-102
1;101;201
1-2;101-106;201;301-303
1;101;201;301
1;101;201
1-4;101-102;201-202;301;401-402;501
1;101;201
1;101-102;201-202
1;101
1;301;401
1-2;101-102;201-202;301
101;301
1;101-102;201-202;301
1;101-102;201-202;301
1;101;201;301
1-2;101
1;101;201;301
1;101-102;201-202;301
101
1-4;101-105;201-204;301-302
1;101-102;201
1-2;101-102;201-202
1;101;201
1;101-102;201-202;301
1-2;101;201;301
1-3;101-102;201-203;301
1-3;101;201
1-2;101-102;201-202;301-302;401-402
1
1;101;201
101-102
1-4;101-102;201-203
1;101
101-102;201-202;301-302
1;101;201-202
1;101;201;301
1;101;201;301
1;101;201
101-102;201-202;301-302
1-4
101-103;201-203;301-302
1-3;101;201;301-302
1;101-102;201-202;301
1;101;201
1;101;201;301
1;101
1;101-102;201-202;301-302;401-402
1;101;201
101-106;201-202
1;201;301
1;101;201
1;101;201-202;301-302
1-2;101-102
1;101-102;201;301
1-7;101-109;201-209;301-309
1;101-102;201-202;301-302
101-106;201-202
1;101-102;201-202;301-302
1;101;201;301
1;101-102
1;101;201
1;101;201
1;101-102;201-202;301
1-4;101-103
1;101-103;201-203;301-304
1;101;201
1;101;201
1;101;201
1;101-102;201-202;301-302
1-2;101-103;201-203;301-302
1;101
1;101-102;201-202;301
1-2;101;201-202;301
1;101-102;201;301
101;201
1;101
1;101
1;101;201
101
1-2
1-2;201;301
1;101;201
1-2;101-102;201-202;301
1;101-102;201-202;301-302
1;101;201
1;101-102;201-202;301-302;401-402
1;101-102;201-202;301-302;401-402
1;101-102;201-202;301-302;401-402
1;101;201;301
1-2;101-103;201-203;301
1;101;201
1-2;101;201;301
1-5;101-105;201-205;301-303
101-103;201;301-302
1;101;201
1;101;201;301
1;101;201
1;101
1;201;301
1-2;101-103;201-203;301-303;401-402
1;5;101;201;301-302;601;701;801;1001;1101
1;101;201;301
1;101;201
1
1;101;201;301
1;101;201;301;401;501;601;701
1-2;101;201
1;101;201
1-6;101-111;201-210;301-313;401-406;501-510;601-606;701-702;801-803;901-903;1000
1-8;101-105;201-205;301-305;401-405
1;101-102;201-202
1;101-102;201
1-2;101-102;201-202;301-302;401
1-2;101-102;201-202;301-302;401
1;101;201;301
1-2;101-102;201-202;301-302
1;101
1;101;201
1;201
101;201
1;101
1-2;101-103
3-14;16-35
1-2;101;201-202
1-2;101;201-202
1;101;201;301
1;201
1;101
101;201;301
101;201
1
101
1-2;101;201-202
101-102
101-102;201-202;301
1;101-103;201-202
1;101;201;301
202
1;101;201
101-102;201-202;301
1-106
1-106
101;201;301
1
1-2;4;101-104;201-204;401-404;501-502
101;201
101-102;201-202;301
1;101;201;301
1;101;201;301
101-102;201-202;301-302
1-2;101-103;201;301
1;101-104;201-204;301
1;101;201
101;201;301
1;101;201;301
1;101;201;301
1;101;201;301
1-2;101-103;201-203
1;101;201-202
1-2;101;201
1;101;201;301
101;201;301
101
1-2;101-102;201
101;201
201;301;401
1;201;301
1;101-102;201-202
101;201;301
101;201
101;201-202
101;201;301
101;201
1;101;201
1;201
1-2;101;201;301
1-2;101-103;201-202
1;101;201;301
1-3;101-103;201-203;301-302
1;101;201-202;301-302
101;301
1;101;201
1;201;301
1;201;301
1;101;201
1;101;201
1
1;101;201
1-4;101-105;201-205;301-302
101-103;201-202
1;101;201
1-2;101;201-202;301
1-3;101;201-202
1;101;201;301
1;101;201;301
1-2;101-102;201
1;101;201
1;201;301
1;101;201;301
1
1;101;201;301
1;201;301
101-102;201;301
1;101;201;301
101;201;301
1;101;201;301
101-102;201-202
1-4;101-102;201-202;301-302
1;101;201;301
101;201;301
101;201;301
1-2;43;101-104;201-204;301-304;401-404;501-504;601-604;701-704;801-804;901-904;1001-1004;1101-1104;1201-1204
1-3;101-105;201-205;301-305
1;101;201
1;101
1
1-3;101-104;201-203;301
1-2;101-102;201;301
1-2;101-102;201;301
1;101;201
1
101-102;201-202;301-302;401-402
312
1;101;201;301
1-2;101;201;301
1-39
1;101;201;301
1;101;201;301
101-103;201-203;301-303;401-403;501-503
101;201;301
1;101-102;201-202;301
1;201
1;101-104;201-204
1;101;201;301
1;101;201;301;401
1;201
1-60
1;101;201
1;101;201
103
1-2;101;201;301
1;101;201;301
1-2;101-102;201-202;301-302;401-402;501-502;601-602;701-702;801
1;101;201;301
1;101-102;201-203;301-303;401-402;501-502;601-602;701-702;801-802
101;201;301
1;101
101-102;201;301-302;401-402;501;601-602;701-702;801
101-102;201;301-302;401-402;501;601-602;701-702;801
1;101;201;301
1;101;201
1-2;101;201;301
1-2;101;201
1;101
1;101;201;301
101
1-2;101;201;301
101;201-202;301-302;401-402
1;101;201;301
1;101;201;301;401
1;101-102;201-202;301;401
1;101;201;301
1;101;201;301
1;101;201;301
1;101;201;301;401
101-102;201-202;301-302
1;101;201
1;101
1;101;201;301
201-202;301-302;401
1-2;101;201;301;401
101;201
1;101;201
101-103;201-203;301-303;401-403;501
101-102;201-202;301-302;401-402;501
1;101;201;301
1;101;201;301
1;101;201
1;101;201;301
1;101;201;301
101;201;301;401
1;101;201
101-102;201-202;301-302
101;201;301;401;501
1-2;101-102;201-202;301
1-3;101-102;201-202;301-302
1;101-103;201-203;301-303
1;101;201
101;201;301
101-102;201-202;301-302
301
101;201;301;401
1;101;201;301;401;501
101-102;201-202;301-302;401-402;501
301-302
101;201;301;401
1;101;201;301
101;201
1-3;12;101-103;201-203;301-303;401-402
1;101-102;201-202
1-2;101;201
1;101;201;301
1-3
1-2;101-102;201;301
1;101;201;301
1;101;201
31
1;101;201;301;401
101-102;201-202;301-303
1;101;201;301
1;101
1;101-102;201-202;301-302;401-402
1;101;201
1;101;201-202;301
1;201;301
101
101;201;301
1;101
101;201;301
101;201;301
101-102;201-203;301-303;401-403
1-4;101-104;201-204;301-304;401-403
1;101-102;201-202
101-102;201-202;301-302;401
1;101;201
1-2;101-104;201-204;301-304
1-5;101-105;201-204;301-306
1;101;201
1;101-102;201-202;301-302
1;101
1-2;101;201
1;101-102;201-202;301-302;401
101-102;201-202;301
1-7;101-102;201-202;301-302
101-102;201-202;301
1-2;101-102;201-202;301-302
1-7;101-102;201-202;301-302
1;101-102;201-202;301-302;401
1-3;101;201;301
1;101;201
1;101;201
1-2;101-103;201;301-302
101-102;201-202;301-302;401-402
1;101;201
1;101;201-202
1;101;201
1-2;101;201
2
1;101;201
1;101;201;301
1-2;101;201;301
1;101;201;301
1;101;201
1;101;201;301
101;201
1;101;201
101;201
1-2;101-102;201-202;301-302
1-3;101-102;201-203;301-303;401
1;101
1;101;201
1;101;201
1;101;201
1;101;201;401
1;101-102
101;301
1;101;301
1;101;201
1;101;201
101;201
1;101;201;301
1-2;101;201
1-3;101-103;201-203;301-302
101-102;201-202
1;101
1-2;101;201;301-302
1;101;201-203;401;501;601
1;101
1-3;101-103;201-203;301-303
1-3;101-103;201-203;301-303
1;101;201;301
1-2;101-102;201-202
1;101;201;301
1;101-103;201-203;301-303;401-403
1-2
1-12
1;201
1;101-102;201-202;301-302
1;101-102;201-202;301-302
1;101;201
1;101;201-202;301-302
1-2;101-102;201-202
1-2;101-102;201-202
101-103;201-203;301-303;401-403;501-503;601-602;701
1;101-102;201-202;301-302;401-402;501;601;701;801
1;101;201;301
1-12
1;101
1;101
1;201;301
1;101;201
1;101;201;301
1;101-102;201-202;301
1-2;101-102;201-202;301-302;401
1;101;201
101;201;301;401
101;201;301;401
1;101;201;301
1;101;201
1-3;101-103;201-203;301-302;401-402
1-3;101-103;201-203;301-302;401-402
1;101;201
1-3;101-103;201-203;301-302;401-402
1;101;301
1;101-102;201-202;301-302;401-402
1;101-102;201;301
1;101;201;301
1
1;101;201;301
201;301
201;301
101-104;201-204;301-304;401-402
101-104;201-204;301-304;401-402
1;101;201-202;301;401
1-2;101-102;201-202
1;101;201
1;101-102;201-202
101;201
1;101;201
1;101;201;301
1;101;201
1;101;201-202;301
1;101;201;301;401
1;201;301
1;101;201
1;101;201
1;101;201
1;101;201
1;101;201;301
1;101;201;301
1
1-3;101-104;201-203;301-303;401-402
1-3;101-104;201-203;301-303;401-402
101-103;201
1-3;101-104;201-203;301-303;401-402
1-2;101-102;201-202;301-302;401-402
1;101;201;301
1;101;201;301
1-2;101;201;301
101;201
1-2;101;201;301
1;101;201
101;201;301
1;101
1-2;101;201
1-2;101;201
1;101-102
1;101;201;301
101;201;301
1;101;201
1;101;201
1;101;201;301;401
1;101;201;301;401
1;101;201
1;101;201;301
1;101-102
1;101;201
1-2;101-103;201-203;301-302
1-2;4;101-104;201-204;401-404;501-502
1-2;11-12;21-22
1-2;101-102;201;301
1;101;201
101;201
1;101;201-202;301-302
1;101-102;201-202;301-302
1;101;201;301
1;101;201-202
1-2;101-102;201-202;301-302;401
1;101-102;201-202;301-302
101;201;301
1;101-102;201-202;301
1;101;201
1;101;201;301
1;201
1-2;101-102
1-2;101-102
101-103
1;101
1;101-102;201-202
1;101-102;201-202
2
1-2;101;201
1
1-2;101-102
101
1;101
1;101;201
101;201
1-2
1-2;11-12;21-22
1-2;11-12;21-22
1-2;11-12;21-22
1;11
1-2;11-12
11;21
11;21
1-2;101;201
1-2;11-19;21-29;31-39
1;11
1;11
1;11
1;11-12;21-22
11
1;11;21
11;21
11-12
1-3;11-12;21-23
1;11;21;31
11-12
11;21;31
1-2;11-12;21
1;11
1;11
1-3;11-13
1;11
11
1;11
11
1;11;21
1;11;21
1;11;21
11;21;31
1-2
1-3;11-13;21-23;31-33;41-42
1-3;11-13;21-23;31-33;41-42
1
1-2;101-102;201-202;301
1;101;201
1-2;101
2
101;201
1;101
1;101
1;101;201
1;101
1;101;201;301
101-102;201-202;301
1-3;101-102
1-2;101-102;201
1;101;201
1-3;201
1-2
1;101-102
1-5
1-5
1-7;101-109;201-205
1;101
1;101
1;101;201;301
1;101;201
1;201-202
1;101;201
1;101;201
1-2;101;201
1;101
1;101;201
1;101;201
1-3;101-102;201;301
101;201;301
1;101-102;201-202;301-302;401-402
1-2;101-102;201-202;301-302;401-402
1;101-103;201-206;301-302;401-406
1-3;101;201
101;201
1;101
1;101
1;101;201
301
1;101;201;301
101;201;301
101;201-206;301-306;401-406;501-506;601-606;701-703;801-803
1-3;101;201;301
1;101;201
1;101
1;101-102;201-202
1;101-102;201;301
1;101;201;301
1-3;101-102;201-202;301
1;101;201
1;101;201-202
1;101;201
1;101;201
1;101;201;301
1;101-104;201-204;301
1;101;201;301
1;101;201
1;101
1;101;201;301
1;101;201
1;101;201;301
1;101;201
1-2;101-102;201-202;301-302
1;101;201-202
1-2
1;101
101;201;301
101-102;201-202;301-302
101;201
1;101-102;201-202;301-302;401-402
101;201
1;101-102;201;301-302
1;101;201
1;201;301
1;201;301
1-2;101-103;201
1;101;201;301
1;201;301
1;101;201;301
1-2;101-104;201-204;301-304;401-402
1;101-106;201-206;301-302
1;101;201
1;101;201;301
1-2;101;201;301
1;201
101;201
1-3;101-103;201-203;301-302
101-102;201-202;301-302
101
1;101;201;301-302
1;101;201;301
1;101;201;301
1-2;101-102;201
101-102;201-202
1-2;101-103;201-203;301-303
101-102;201-202
1;101;201
1;101;201
101-102;201
1;101;201
1;101
1;101
1;101;201;301
101;201;301
1;101-102;201;301
1-3;101
1;101
1-2;101-103;201-203;301
1;101-102;201-202
1-4
1;101;201
1;101
1;101
201;301;401-402
201;301;401-402
1-2;101;201
1;101
1;101-102;201-202;301
1;201
1;101;201;301
1-2;101-103;201-203;301-302
101-102;301
1-2;101-102;201-202;301-302
1-2;101-103;201-202;301
2-6
1;101;201
1-6;11;13-16;21-26
101
1-2;101-102;201-202
1-2
1-2
1;101
1;101
101-103;201-203
101-103;201-203
1-2;101-102;201-202
101-102;201-202
101-103;202
1-2;101-102;201-202;301-302
1-2;101-102;201-202;301
1
1
101-102;202
1-2
1-3;101-102
101
101;201
101;201
1-2;101-102;201-202;301-302
1-2;101-102;201-202;301-302
1-4;101-102;201-203
1-2;101-102;201
1-2;101-102
1;101;201
1;101-102
1
101;201-204;301-304
1-2;101-104;201-203;301-303
1-6;101-106;201-204
1-4;101-105;201-202
1;2;101;102
1;2;101
1;101-103
1-4;101-105;201-202
1;101
101;201;301
1;101
1-2
1;101
101-103;201-202;301-303;401-402
1;101
1-2
101-105;201-203
101;201
1;101;201
1;101;201
101;201
101;201;301
1-2
1;101;201
1;101;201;301
1-2;101;201
1;101;201
1;101
1;101-102;201-202
1;101
1-3;101;201
1;101;201
1-2
1-2
1;101;201;301
1-4
1-6
1-2;101-103;201-203;301-303
101-106;201-206;301-306;401-406
1-2;101-102;201-202
1-2;101-102;201-202
1-2;101-102;201-202
1-2;101-102;201-202
1-2;101-103;201-203
1;101-102;201-202
1-5
1-3;101-102;201
1;101;201
1;101;201
1;101;201;301
1;101-103;201-203;301-303
1;101-102;201-202
1;101-103
1;101;201
1;101-102;201
1-2;101-102
1;101;201
1;101;201
1-2;101-102;201-202
1-2;101-102;201-202;301-302
1-2;101-102;201-202
1-3;101-103;201-203
1-2;101-102;201-202
1-2;101-102
0001;0002;0003;0101;0102;0103;0201;0202;0203
0001;0002;0003;0101;0102;0103;0201;0202;0203
0001;0002;0003;0101;0102;0103;0201;0202;0203
0001;0002;0003;0101;0102;0103;0201;0202;0203
1-2;101-102;201-202;301-302
1-2;101-102;201-202;301-302
101;201
101;201;301
1;101
101-102;201
1;101
1-2;101-102;201-202
101;201-202
101;301
1;101
1;101;201
1;101;201
1;101
101;201
1;101;201-202;301-302
1;101
1;101
101-102
1;101-103;201-202;301-302
1;101
101
1-2;101;201;301
1;101
1;101
1;101
101;201;301-302;401
101
1-2
1;101;201
101;201;301-302
1-2;201
1;101-103;201
1;101
101;201
1-3
1-2;101-102;201
1-2;101-102;201-202
1;101
1;101;201
1-2;101-102
101-104;201-204
1;101-103;201
1;101;201;301;401
1;101
1;101;201-202;301-302
1-2;101-102;201-202;301-302
101-105;201-205
1-2;101-102;201-202
1-2;101-102;201-202;301-302
1-2;101-102;201-202;301
1-2;101-104;201
1;101
1;101;201
1-3;101-105;201-205;301-305
1-3;101-104;201-204;301-303
1;101
1;101
1;101
1;101;301
101
1;101-102;201
1;101;201-203
1-13
101-103;201-203;301-303;401-402;501-502
1;101
1;101
1;101;201
101;201
101-102;201
1;101-102;201-202;301
1;101;201
1;101
1;101
1-2;101-102;201-202;301
1;101;201;301
1-2;101-103;201
101-102;201-202;301-302
1;101
1;101;201
1;101
1;101
1;101;201
1;101
1;101
101-102
1;101;201
101;201-202
1;101;201
101;201
101;201
1-2;101-102;201-202
1
1;101;201;301
2
1;101;201
1;101;201;301;401
1;201;301
1;101
1-2;101;201
1;101;201;301
1;101
1;101
101-102;201-202;301-302
1-3;101
1;101
1;101-102;201
1;101
1;101;201
1;101
1;101
1-2;101-102;201-202
101;201;301
1-3;101-103;201-203
1-3;101-103;201-203
1-3;101-103;201-203
1-3;101-103;201-203
1-3;101-103;201-203
1-3;101-103;201-203
1-3;101-103;201-203
1-3;101-103;201-203
101-102;201
1-5;101-105;201-204
1;101
1;101;201;301
101;201;301-302
1;101-102;201-202;301-302
1-2;101-102;201-202;301-302
1-3;101-102;201-202
1;101;201
1;101;201
1;101;201
1-2;101
101
1;101;201;301
1;101;201;301
1-2;101;201-202;301-302
1;101
1-3;101-104;201-204;301-304
1;101
1-2;101-102;201
1-4;101-104;201-204
1-4;101-104;201-204
1-2;101-105;201-204;301-302
1-2;101-105;201-204;301-302
1;101
101-103;201-203;301-302
1-5;101-102;201-202
1;101;201
1-3;101-103;201-203
1-5;101-102
1;101-102
101-102;201-202;301-302
1;101-102;201-202;301
1
101;201;301
1-23
1-3;101-102;201-202;301-302
1-6;101-111;201-210;301-313;401-406;501-510;601-606;701-702;801-803;901-903;1000-1001
1;101;201;301
1;101;201
1;101;201;301
1-5;101-103;201-206;301-302
1-2;101-102;201-204;301-304;401-403;501-504
1;101
```
</details>

**Comment by Adamant36** [comment]
> I don't know where you get the number of 99% good tagging. If I look around, I see at least 60% long lists.

Look through the pages for tags usage on TagInfo. It takes like 4 or pages to get to one like your example when its already in the low usage numbers. Then it rarely comes up. Like the first one that comes up, 
1-2;101;201, is 21 pages in and only has 77 uses. After that they are extremely infrequent and the usage is to low to matter. Its like less then 1% of the total. Just the page which doesn't contain the problem your talking about is like 40,000 uses on its own. My guess is the bad ones are a couple hundred at best. Which is less then 1% of the total usage. Even if there are  1,000 it still isn't 1% of the total (it's like 0.005), but I highly doubt it is anywhere near that. 500 or so out of 185,835 isn't anything. You could re-tag them in a weekend, or just ignore them and accept that nothing is perfect.

**Comment by Adamant36** [comment]
Maybe your region is an outlier because it doesn't seem to represent the usage of the tag overall, as stated above. Id suggest fixing the wrong ones that you can. I don't see the problem with that.

**Comment by sanderd17** [comment]
> You could re-tag them in a weekend

Just a question I asked before, to what should you retag those values? They are correct according to the wiki.

It's the job of the renderer to give a good visualisation of the map. It's not the job of the mapper to map for the renderer? Or do you claim otherwise?

**Comment by Adamant36** [comment]
> Just a question, to what should you retag those values?

In your first example since they do seem to be a range of numbers, just with some missing, I'd re-tag it as four separate addr:flats nodes from left to right. Starting with 0.1-0.2, then 1.1-1.3, 2.1-2.2 etc etc. That seems correct to me. Likely it was four separate areas of the building with different flat numbers that the just combined because they were lazy. Which isn't how the tag is meant to be used IMO.

> It's the job of the renderer to give a good visualisation of the map. It's not the job of the mapper to map for the renderer? Or do you claim otherwise?

My claim, and I'm pretty sure everyone will agree with me, is that rendering or not rendering something shouldn't be based on an extremely small regional sample of outliers in tagging that can easily be fixed as I've outlined above. 

> And yes, it might be a problem in our region. We have many small apartments (just 3 or 4 floors), with a few flats per floor, and usually number them by floor

Then it sounds like my suggestion to tag them as separate nodes above is a sound one. Since it sounds like they are different objects. Maybe in conjunction with the level tag if you have that information, or come up with something like a flats:floor tag. I'm sure it would be useful outside of just your examples. Also, you should discuss it with your local community. One of the great things about OSM is the ability to come up with novel ways to deal with exactly these types of situations and there's a lot of grey area between render/don't render something. 

> There's really no space to render that decently, and at the moment, mapping the addr:flats obscures many other elements (like housenumbers) which are a lot more important.

Your really working backward from own conclusion here. You can't say a tag with 46 thousand uses isn't important and every blocks everything in some situations. Again, it's not an either or thing though. Correct the ones that you can correct how I suggested, discuss it with your local community, find a novel way to remedy it, and then reevaluate things.

**Comment by Zverik** [comment]
The idea for `addr:flats` was to be tagged on building entrances — to specify which flats an entrance leads to. This indeed messes things up when rendered on bulidings or other polygons.

Of course that doesn't solve the case of addressed entrances. Lists of numbers on the opening pictures should go into `addr:housenumber`. This has usually been solved by tagging housenumbers on entrances (see Brussels).

**Comment by Adamant36** [comment]
> This has usually been solved by tagging housenumbers on entrances

Tagging things that are properties of the whole building on the entrance has seemed really weird to me for some reason. Like nesting tags or something. The address applies to the whole building. Not just the front door. Plus, you have to know where the entrance is in the first place. Maybe I could sorta see it for addr:flats since "flats" isn't a single object to tag it on, but it makes no sense for addr:housenumber IMO. So, I think it could go either way in the first example. Mapping it on the entrance might actually make the rendering worse. I think my suggestion of splitting them up by range is probably the more sound option.

**Comment by jeisenbe** [comment]
Also see #4155

**Comment by andrewharvey** [comment]
I agree with @sanderd17 `addr:flats` should not be rendered as a text label because it'll often be quite long. The wiki which matches current mapping practice says:

> Use addr:flats=* for marking range of numbers of flats behind a door, for example, addr:flats=1-20. If there is gap in numeration, mark numbers and intervals with semicolon: addr:flats=3-7;10;14;16-18.

An example is `addr:flats=101-103;201-208;301-306;401-405;408-409;501-507;601-605;701-704;801-806;901-904;1001-1005;1101-1108`. Too long to render, but 100% correct tagging. I have scenarios where the list exceeds the 255 tag value limit and have to split the values across addr:flats, addr:flats1, addr:flats2 etc.

Here is another example of a real world value at https://www.domain.com.au/building-profile/668-bourke-street-melbourne-vic-3000

`1-3;10-12;100-103;110-112;200-203;210-212;300-303;305-312;400-403;405-412;500-503;505-512;600-603;605-612;700-703;705-712;800-803;805-812;900;902-903;905-911;1000;1002-1003;1005-1011;1100;1102-1103;1105-1111;1200;1202-1203;1205-1211;1300;1302-1303;1305-1311;1400;1402-1403;1405-1411;1500;1502-1503;1505-1511;1600;1602-1603;1605-1611;1700;1702-1703;1705-1711;1800;1802-1803;1805-1811;1900;1902-1903;1905-1911;2000;2002-2003;2005-2011;2100;2102-2103;2105-2111;2200;2202-2203;2205-2211;2300;2302-2303;2305-2311;2400;2402-2403;2405-2411;2500;2502-2503;2505-2511;2600;2602-2603;2605;2607;2609-2611;2700;2702-2703;2705;2707;2709-2711;2800;2802-2803;2805;2809-2811;2900;2902-2903;2905;2907;2909-2911;3003;3006;3009;3012-3013`

Yep that's a 30 story building where they don't like the number 4, which forces these ranges to be even longer.

According to https://taginfo.openstreetmap.org/keys/addr%3Aflats#combinations 20% (73k) of these tags appear together with building=* (so not on the entrance), 50% (181k) appear with entrance=* tag (so on a building entrance). So it appears that people are using the tag on buildings.

**Comment by andrewharvey** [comment]
> Then it sounds like my suggestion to tag them as separate nodes above is a sound one. Since it sounds like they are different objects. Maybe in conjunction with the level tag if you have that information, or come up with something like a flats:floor tag. I'm sure it would be useful outside of just your examples. Also, you should discuss it with your local community. One of the great things about OSM is the ability to come up with novel ways to deal with exactly these types of situations and there's a lot of grey area between render/don't render something.

An apartment building with 400 units we shouldn't be having 400 nodes all on top of each other, that would be an editing nightmare. Using addr:flats allows all 400 different units to be recorded in OSM with a single building polygon and address tags.

**Comment by Adamant36** [comment]
> Yep that's a 30 story building

I imagine you could likely (and it would probably be correct to) add the ranges of flat numbers as an addition to something like building:levels (or whatever the proper tagging for multi-story buildings is) areas. Realistically though no one is going map a thirty story building as thirty different anything though and trying to add every flat number for such buildings is clearly "mapping every blade of grass." Plus, really, no one cares about what flat is where in the building at that point either. 

Also, I'm sure we could find a single ridiculous edge case to poopoo any  rendering any tag. I don't think it's valid way to approach to this though if there's almost zero chance of such things being mapped in any amount (or at all) IRL. Even your shorter example is extremely unlikely to occur and if it did there's no reason they couldn't just be split into a couple of different nodes. Since doing so wouldn't technically be wrong either. That said, the longest actual example I could find is `1;10;11;12;13;14;15;16;17;18;19;19A;2;20;21;22;23;24;25;26;3;4;5;6;7;8;9`. Which could easily be split up somehow. The vast majority are no more then 3 numbers though. Most are 2.

**Comment by andrewharvey** [comment]
> I imagine you could likely (and it would probably be correct to) add the ranges of flat numbers as an addition to something like building:level (or whatever the proper tagging for multi-story buildings is) areas.

I don't understand what you mean. The proper tagging for a multi-story building is `building=* + building:levels>1`.

> Realistically though no one is going map a thirty story building as thirty different anything though

If someone was to do indoor mapping they might map each floor, but I'm not talking about that, I'm just talking about mapping the building itself as `building=apartments + building:levels=30 + addr:housenumber=N + addr:flats=Y`.

> And trying to add every flat number for such buildings is clearly "mapping every blade of grass."

I disagree.

> Plus, really, no one cares about what flat is where in the building at that point either.

Exactly, which is why I wouldn't and most people wouldn't map each addr:unit=* as it's own node within the building, I'd just map addr:flats=* as the list of units within the building.

> Also, I'm sure we could find a single ridiculous edge case to poopoo any rendering any tag.

It's not a single edge case, it's very common, and it's not ridiculous.

>  I don't think it's valid way to approach to this though if there's almost zero chance of such things being mapped in any amount (or at all) IRL. 

But people are mapping addr:flats this way, so as a style you can't change how people are mapping or tagging in OSM you just have to accommodate.

> Even your shorter example is extremely unlikely to occur 

I disagree, they are common where you have apartment buildings where the numbers arent just 1-N but per floor. eg units on level 1 would be 101, 102, 103, etc. then on level 2 201, 202 etc. So you need to represent this as 101-103;201-203;etc which for high rise apartments creates a long list.

> if it did there's no reason they couldn't just be split into a couple of different nodes. Since doing so wouldn't technically be wrong either.

If they are all at the same building, then they should be as part of the single building and part of the address.

**Comment by Adamant36** [comment]
> I don't understand what you mean. The proper tagging for a multi-story building is building=* + building:levels>1.

I just didn't have access to the Wiki at the time and I couldn't remember if that was correct tag for building levels or not. Also, I was thinking about 3D mapping. Which I think is the only realistic (although still stretching it) scenario where something like that could be mapped in connection with the flat numbers. Otherwise, the flat numbers are extremely useless. 

>  I'm just talking about mapping the building itself as building=apartments + building:levels=30 + addr:housenumber=N + addr:flats=Y

Let me ask you this, how is having specific flat numbers on a single area at all useful or useable at that point? Like, if the person is looking for flat number whatever in a thirty story building they already know that the thirty story building has that flat somewhere in it. Adding a fifty number string of flat numbers to a thirty story building area doesn't help anyone actually find anything though. Let alone does it convey anything useful. Again, they already know flat "whatever" is in the building somewhere. 

> I disagree.

See my last response. Really, OSM is suppose to be a map of things that exist on the ground and can be confirmed. Not everything that exists and can be confirmed though. Just useful, actionable things. To me a thirty story building with a bunch of flat numbers attached to a single building area is neither. Just like "mapping every blade of grass" isn't. Just because you can do something, doesn't mean you should. 

> It's not a single edge case, it's very common, and it's not ridiculous.

You didn't provide a link to a real world example. So I assumed it was either an edge case or didn't exist in the first place. Also, I researched it myself and couldn't find any examples even close to that many numbers. Not to say they don't exist, but seriously doubt it's that many. 

Look it at this way, there's 383,992 uses of addr:flats. If there's like 200 examples, which seems a little high to me, then it's still only 0.0001% of the total usage.  That's nothing. Literally every tag out there has that many "bad usage" examples. Probably it's more like only 4 or 5 "bad" examples. Say it isn't and there's multiples more, there's really no hard line anyway. Otherwise, nothing would be rendered. 

> I disagree, they are common where you have apartment buildings where the numbers arent just 1-N but per floor.

So? From what I can tell they aren't being mapped if they do exist. Which should be the important thing, because if your going to argue something shouldn't be rendered because such and such issue, then it should actually be an existing issue at the time or at least somewhat likely to occur. Again though, I doubt they are common or there would be a more then nominal amount of them currently mapped. When there just isn't. 

> If they are all at the same building, then they should be as part of the single building and part of the address.

I disagree. I see nothing wrong with mapping things as separate nodes if they are on separate floors or whatever. I'm pretty sure that's how things are already done with underground features. As well as places where a shop is on the bottom floor of a multiple story building and the "apartments" above it have different addresses. Otherwise, you'd be adding the address for the apartments to the stores POI. Which would be stupid and isn't possible anyway.

**Comment by andrewharvey** [comment]
> Let me ask you this, how is having specific flat numbers on a single area at all useful or useable at that point?

Why does that matter? As OSM mappers we try to model the real world, it doesn't matter if it's useful to me or not I still map it as one day it might be useful to someone.

For geocoding it provides validation, if I search for a specific flat number at a specific address, with addr:flats OSM will be able to tell me if that flat number exists at the address or not.

> Again, they already know flat "whatever" is in the building somewhere.

How do they know it exists or not if it's not mapped? Maybe someone wants to generate a list of all addresses at the building, to do this they need addr:flats.

Or maybe that 30 story building is actually comprised of two towers each with the same address, or maybe it's one tower but half the units are from one entrance and the other from another entrance. You need to list out all the units for each entrance.

> You didn't provide a link to a real world example. So I assumed it was either an edge case or didn't exist in the first place.

I'm working on an address import at the moment which would import many long addr:flats. See https://gitlab.com/alantgeo/vicmap2osm/#removing-duplicates

> From what I can tell they aren't being mapped if they do exist.

Well then I better go outside and start mapping some :wink: 

>  I see nothing wrong with mapping things as separate nodes if they are on separate floors or whatever.

Sure if you want to do full indoor mapping of a building then yes, but as an intermediate step for someone wanting to add more detail to the building without doing indoor mapping, then addr:flats is much easier to manage than just 300 nodes on top of each other in the building centroid.

**Comment by pnorman** [comment]
We've drifted a long way from discussing how to improve the rendering of addr:flats.

**Comment by Adamant36** [comment]
> We've drifted a long way from discussing how to improve the rendering of addr:flats.

I don't think so. At least from my side. My comment was aimed at accomplishing a few things, 1. Getting an example from the map of the extremely long string of flat numbers that @andrewharvey says exists 2. Arriving at understanding of how much of a problem such tagging actually is in the real world. 

I'd say both are perfectly relevant and on topic. Even if @andrewharvey was unable to aid me in accomplishing either. 

At the end of the day if he can't then I'd say there's zero reason not to render this. At least not based on the `1-3;10-12;100-103;110-112;200-203;210-212;300-303;305-312;400-403;405-412;500-503;505-512;600-603;605-612;700-703;705-712;800-803;805-812;900;902-903;905-911;1000;1002-1003;1005-1011;1100;1102-1103;1105-1111;1200;1202-1203;1205-1211;1300;1302-1303;1305-1311;1400;1402-1403;1405-1411;1500;1502-1503;1505-1511;1600;1602-1603;1605-1611;1700;1702-1703;1705-1711;1800;1802-1803;1805-1811;1900;1902-1903;1905-1911;2000;2002-2003;2005-2011;2100;2102-2103;2105-2111;2200;2202-2203;2205-2211;2300;2302-2303;2305-2311;2400;2402-2403;2405-2411;2500;2502-2503;2505-2511;2600;2602-2603;2605;2607;2609-2611;2700;2702-2703;2705;2707;2709-2711;2800;2802-2803;2805;2809-2811;2900;2902-2903;2905;2907;2909-2911;3003;3006;3009;3012-3013` example.

**Comment by imagico** [comment]
I concur with @pnorman - recent discussion is fairly non-productive.  As said plenty of times in the past this is not the place to discuss how tags should be used, we need to look at how tags are actually used. It is fairly evident that there are uses of `addr:flats` where the current display style is counterproductive for map readability.  OTOH there are a lot of uses of the tag where rendering it in some form (not necessarily in the current simple concatenation with the housenumber) would be beneficial for both mapper feedback and map usefulness.

But until someone takes a planet file and does an analysis of the length distribution of values of `addr:flats` any discussion how to address this problem is pretty pointless.

**Comment by andrewharvey** [comment]
> Even if @andrewharvey was unable to aid me in accomplishing either. At the end of the day if he can't then I'd say there's zero reason not to render this.

I went for a short walk this morning intent on adding a bunch of addr:flats tags, many older apartments used sequential numbers 1-N which was easy and short to tag. But newer apartments particularly those marketed to Chinese investors had the kinds of values I described, they'd be long even without blacklisting the number 4, but with, one here I checked this morning I had to overflow the 255 tag value limit twice to split the value across `addr:flats`, `addr:flats1`, `addr:flats2.

Not representitive at all for a global database, but just a data point of real world examples of longer values.

https://www.openstreetmap.org/way/466884766
https://www.openstreetmap.org/way/504331827

I think it's very clear from all this that longer values are valid how the tag is documented.

I'm now generating a report from the planet dump to see the length distribution of existing tagged values.

PS. I think we could do with a bit less intimidation in this thread, all it does is drive away contributors and people trying to help.

**Comment by andrewharvey** [comment]
At the moment in OSM 93% of `addr:flats` are 10 characters or less, 20k of 310k total tags are > 10 characters.

Source code/data below.

```
wget -O - 'https://ftp5.gwdg.de/pub/misc/openstreetmap/planet.openstreetmap.org/pbf/planet-latest.osm.pbf' | osmium tags-filter --input-format=pbf --output-format=pbf --omit-referenced - 'addr:flats' > flats.osm.pbf
osmium export --output-format=geojson --omit-rs flats.osm.pbf > flats.geojson
ogr2ogr -select 'addr:flats' -f CSV flats.csv flats.geojson
```

```
wc -l flats.csv
309616 flats.csv
```

```bash
#!/bin/bash

while read line; do
    ((histogram[${#line}]++))
done < flats.csv

echo "Length Occurrence"
for length in "${!histogram[@]}"; do
    printf "%-6s %s\n" "${length}" "${histogram[$length]}"
done
```

![2021-06-26_20-34](https://user-images.githubusercontent.com/117278/123510286-e12e4100-d6bd-11eb-9af7-81fb9f7d4104.png)

```
Length Occurrence
1      8554
2      2359
3      84223
4      43313
5      88110
6      14716
7      37652
8      1480
9      6360
10     2231
11     1539
12     909
13     3589
14     1511
15     675
16     870
17     1919
18     252
19     989
20     222
21     781
22     294
23     246
24     284
25     1372
26     104
27     117
28     375
29     295
30     57
31     223
32     120
33     296
34     97
35     69
36     78
37     305
38     27
39     37
40     121
41     153
42     39
43     88
44     30
45     88
46     78
47     17
48     30
49     218
50     20
51     45
52     78
53     51
54     21
55     54
56     28
57     66
58     32
59     21
60     25
61     78
62     26
63     14
64     50
65     57
66     15
67     28
68     10
69     27
70     25
71     15
72     16
73     76
74     9
75     14
76     25
77     16
78     14
79     26
80     9
81     35
82     18
83     10
84     10
85     46
86     6
87     8
88     14
89     15
90     8
91     21
92     6
93     21
94     15
95     10
96     7
97     34
98     2
99     12
100    9
101    7
102    6
103    15
104    8
105    15
106    15
107    12
108    5
109    32
110    9
111    12
112    9
113    16
114    7
115    13
116    4
117    9
118    10
119    4
120    3
121    22
122    6
123    7
124    10
125    7
126    8
127    12
128    4
129    10
130    9
131    9
132    5
133    18
134    6
135    4
136    11
137    3
138    5
139    7
140    3
141    7
142    10
143    3
144    4
145    16
147    4
148    12
149    11
150    5
151    20
152    2
153    5
154    9
155    4
156    6
157    11
158    1
159    4
160    6
161    7
162    11
163    5
164    2
165    2
166    5
168    1
169    6
170    3
171    4
172    5
173    5
174    2
175    3
176    3
177    3
178    4
179    3
180    3
181    9
182    3
183    4
184    8
185    2
186    6
187    4
188    1
189    3
190    8
191    2
192    5
193    8
196    2
197    2
198    3
199    3
201    5
202    3
203    2
204    3
205    4
206    2
207    7
208    6
209    1
210    2
211    6
212    1
213    3
214    6
215    1
216    1
217    4
218    4
219    2
220    2
221    1
222    4
223    5
224    3
225    3
227    2
229    1
230    1
231    4
232    6
233    3
235    2
238    7
240    1
241    2
244    1
245    1
246    2
247    1
248    1
249    4
251    2
252    1
253    3
254    1
255    6
256    5
257    1
```

**Comment by andrewharvey** [comment]
I'm not sure how feasible it is to implement, but probably the best solution is anything <= 7 characters (7 characters captures something like `101-909`) can be rendered, but anything longer should not render. This captures the vast majority of cases which can be rendered, and by not showing longer values it lessens the impact of people deleting still valid tags just because it makes the map rendering ugly.

**Comment by andrewharvey** [comment]
Completely untested but proposed fix at #4430

**Comment by imagico** [comment]
Thanks, that is very useful.

My suggestion would be to try:

* separating the rendering of addr:flats and addr:unit from addr:housenumber in rendering and prioritizing addr:housenumber (and this way avoiding that addr:flats and addr:unit of one address block the more significant addr:housenumber of another one).
* distinguish rendering of addr:flats and addr:unit from addr:housenumber in text styling (it is currently ambiguous which tags are actually rendered)
* apply some shortening to longer addr:flats values (>7 characters would be a clear cutoff based on the current distribution) - like turning `A;B;C;D;E` into `A...` or `A;B...`.  Since most longer addr:flats values are either repetitive/overlapping for different neighboring addresses (several appartment blocks with different housenumbers but with same or very similar lists of flats) or roughly consecutive and non-overlapping (different blocks from the same housenumber) such a shortening would provide some useful info to the map reader in compact form without taking a lot of space.

I don't think just dropping rendering of longer values is a good idea - that is potentially confusing for mappers.

**Comment by andrewharvey** [comment]
> separating the rendering of addr:flats and addr:unit from addr:housenumber in rendering and prioritizing addr:housenumber

partly related to #3568, but in Australia the addr:unit and addr:housenumber would go together as `{addr:unit}/{addr:housenumber}`.

addr:unit shouldn't be used together with addr:flats, so if there is addr:flats instead then `{addr:flats}/{addr:housenumber}` works too.

> distinguish rendering of addr:flats and addr:unit from addr:housenumber in text styling (it is currently ambiguous which tags are actually rendered)

Again I can only comment on the Australian localisation, but this should be unit/number, no need for different styles. For my region the current number space unit is very confusing.

> apply some shortening to longer addr:flats values

I guess we could do that. At first it might be confusing to mappers, if you see "..." then you think that was a literal tag value and might mean the tag needs fixing only to find it was just a rendering style.

If the tag value is long then really there's not much benefit in showing the first 7 characters as they probably aren't as useful as if the tag value was <= 7 chars.

> A;B;C;D;E

I don't think we should try to cater for this, since the wiki specifically says you can use a range to compress the values eg A-E which would fit within the 7 characters.

> I don't think just dropping rendering of longer values is a good idea - that is potentially confusing for mappers.

Yes, but also rendering them or trimming with ... is also potentially confusing to mappers. So I think dropping the values from display is probably the best compromise.

**Comment by imagico** [comment]
Keep in mind that we are creating an international map style that means to show mapping under different cultural conventions with equal determination.  And we also aim to render incomplete mapping (some address components being mapped while others are missing) in a transparent and non-confusing fashion.

If abbreviating address tags with an ellipsis could be misinterpreted as being literally tagged as such using a vertical/midline ellipsis can probably help - see https://en.wikipedia.org/wiki/Ellipsis#Computer_representations.

**Comment by dch0ph** [comment]
The PR in development, #5272, addresses shortening of `addr:housenumber` and `addr:flats` where these contain lists of values, but this would leave unaddressed the problem of how to render `housenumber` and `flats`/`unit` where both are present - currently they are just jammed together `{housenumber} {unit/flats}`, which is very unsatisfactory.

There are a variety of regional styles, e.g. "Flat 5, 16 Foo Road" (UK), "5/16 Foo Road" (Australia?), but these won't work for a global map.

Some thoughts:
- If an address label contains flat/unit information, then the flat/unit is the more important piece of information to identify that particular entrance / building. (We are not trying to display addresses.)
- This is different to the priority of the _label_ for rendering. Labels containing flat/unit detail are less important than those just containing the overall housenumber. This is partially recognised currently with flat/unit only being rendered from Z18 (rather than Z17).

How about `{flat/unit} [{housenumber}]`, where [] indicates the presence of additional, but secondary, information?

e..g. `5 [16]` `3 [16]` on different units, or `1-5 [16]` for a set of flat numbers? 

This may be a cultural thing, but [] feels more "neutral" than ().

I don't that think dropping `housenumber` if flat/unit information is present is viable, otherwise you would identical-looking labels showing different information, `5` (unit number) next to `16` (housenumber).

It should be straightforward to prioritise labels without flat/unit information by a SORT BY in SQL.

**Comment by imagico** [comment]
> How about {flat/unit} [{housenumber}], where [] indicates the presence of additional, but secondary, information?

Do you suggest this because you think this is the most intuitive design approach or because it is technically easier to implement than different designs for housenumbers and flats/unit without punctuation characters?

In particular keep in mind that this needs to work not only with Arabic numerals/Latin characters but with any kind of writing system in use.  And also keep in mind that currently we do not only concatenate housenumbers and flats/unit, but also housename.

Also to consider here: #3038, #4687.

> It should be straightforward to prioritise labels without flat/unit information by a SORT BY in SQL.

Yes, but my thought was to prioritize the housenumbers relative to the flats/unit.  Meaning if there is little space the housenumbers always show (even when flats/units are mapped as well) and the flats/units would only show if they don't interfere with the housenumbers.

The side effect of such an approach would be that the flats/unit in some cases might show without the corresponding housenumbers (like - for example - the housenumber is blocked by some symbol while the flats/unit label is not).  That is because in Mapnik you currently cannot make placement of a symbol/label depend on the successful placement of a different symbol/label already placed.

**Comment by dch0ph** [comment]
> Do you suggest this because you think this is the most intuitive design approach or because it is technically easier to implement than different designs for housenumbers and flats/unit without punctuation characters?

Within the current framework, I don't see much option but to use punctuation to indicate "supporting information". Given that the resolution on the OSMF deployment is at the margins of legibility (#2265), I don't think using a smaller typeface size is viable.

Looking at your proposals for address/entrance nodes, I've mocked up a modified proposal, which might be workable:

<img width="207" height="143" alt="Image" src="https://github.com/user-attachments/assets/8719c25f-93ad-4efa-be71-6faa947c870d" />

- Distinguish between address labels with flats/unit/ref tags (Z18+) and those without (Z17+). 
- The former don't show `housename` if present, the latter do.
- But,  in contrast, I've highlighted the flats/unit/ref value and put any housenumber in a pseudo-subscript (but without shrinking the typeface size for legibility). This makes sense IMHO because the flats/unit/ref tag is the most important on the particular entrance/address node.
- Allow `ref` on entrances, using italics to differentiate from flats/unit (consistent with our use of italics for `ref` elsewhere)

Of course, the main problem is that we can't implement non-uniform labels in CartoCSS, and would have to split labels containing both flats/unit/ref and housenumber into two labels, with strong possibility that one will end up being blocked, leading to a confusing result.
These problems would be less severe if the style were to be deployed to Z20, but I can't see that ever happening (because scrapers).

So, yes, `1 [10]` etc. is clunkier and less intuitive.

Not rendering `housename` when flats/unit/ref taggging is present will help with areas like industrial estates:

<img width="124" height="99" alt="Image" src="https://github.com/user-attachments/assets/d25a7ec3-9e10-462a-93af-d403552d012a" />

Here, a `craft=window_construction` POI is not being rendered, and to make the clutter worse, a long `addr:housename` is shown instead from the address tagging.

**Comment by imagico** [comment]
> Distinguish between address labels with flats/unit/ref tags (Z18+) and those without (Z17+).

Does that mean you want to show an `addr:housenumber` tagged alone from z17 but only show it from z18 if also `addr:flats` is tagged?  That would be not very intuitive.  Showing flats/unit/ref later than `addr:housenumber` makes sense of course.

> But, in contrast, I've highlighted the flats/unit/ref value and put any housenumber in a pseudo-subscript (but without shrinking the typeface size for legibility). This makes sense IMHO because the flats/unit/ref tag is the most important on the particular entrance/address node.

Using a systematic relative positioning is a viable way to differentiate between different labels.  But you would need to add a reference marker in case certain tags are missing.  I showed that [here](https://imagico.de/blog/en/addresses-and-entrances/).  The reference marker would also help in case some labels are blocked.

> Not rendering housename when flats/unit/ref taggging is present will help with areas like industrial estates: [...]

That example is just nonsensical tagging.

I don't see any indication of `addr:housename` and `addr:unit` being mutually exclusive.  The problem here is de-duplication.  Several addresses with different `addr:unit` will often share the same `addr:housename` and displaying `addr:housename` for each of them will be wasteful.

What would be nice to have is displaying housenumbers on POIs when there is space for that.  But that would also result in a lot of clutter without de-duplication.

It might also be a good idea to more clearly distinguish address labels from other labels in general.  In particular the semantic difference between `addr:housename` and building names (`name=*` on `building=*`) would become much clearer with that.

**Comment by dch0ph** [comment]
> Does that mean you want to show an `addr:housenumber` tagged alone from z17 but only show it from z18 if also `addr:flats` is tagged? That would be not very intuitive. Showing flats/unit/ref later than `addr:housenumber` makes sense of course.

The current rendering is not very intuitive when both entrances and buildings are tagged, since the housenumber appears duplicated at Z17 and then the key information distinguishing the unit/flats entrances appears at Z18 tucked on the end.

Let me refine the proposal to:
- Entrances are only labelled from Z18. 
- Entrances only show `housename` if none of number/ref/flats/unit are present.

I don't see a problem with differentiating labelling of entrances vs address nodes/buildings. The key information for entrances is "what entrance is this". 

I then wouldn't be able to dodge the question of what do when all of flats/unit + number + name are present. In the "systematic relative positioning" (as opposed to using `[]`), you would then need a 3 part label. 

> Using a systematic relative positioning is a viable way to differentiate between different labels. But you would need to add a reference marker in case certain tags are missing. I showed that [here](https://imagico.de/blog/en/addresses-and-entrances/).

Possibly, but missing information doesn't arise in the relatively simple scheme: flats/unit vs. housenumber are shown normally if only one is present.

> I don't see any indication of `addr:housename` and `addr:unit` being mutually exclusive.

But I'm pretty sure that the semantics are different. A "normal" house name has the same function as `addr:housenumber`, i.e. strictly only one of number/name is needed, and there is no duplication problem. With `addr:unit`, the `addr:housename` is something like "Block A" or (even things like) "X Business Park" (which is pretty common in the data). Hence `addr:unit` on an entrance doesn't need `addr:housename` (which should be tagged on the building). Either way, I feel that `housename` is pretty "disposable" on entrances.

> The problem here is de-duplication.

Agreed. The addressing tagging encourages duplication. But I think we can simplify the problem a little by distinguishing entrance labelling.

**Comment by imagico** [comment]
> Entrances are only labelled from Z18.

That would mean discouraging tagging addresses on entrances vs. tagging on buildings/address nodes.  

> Entrances only show housename if none of number/ref/flats/unit are present.

I know i also don't show housename on entrance nodes in my demo, but this still seems a bit like pseudo-deduplication.  You skip displaying `addr:housename` on the entrance because you expect it already being shown based on being tagged on the building or an address node.  I am not sure this is a good idea.  Tagging addresses only on entrance nodes is widely considered perfectly valid mapping and, of course, these entrance based addresses can have an `addr:housename`.  And address tags on a building are often blocked by POI symbols.

I see that we are a bit in Yak shaving mode here, coming from shortening lengthy number lists to differentiating different address components and from there now to the problem of de-duplication.  So please don't consider my remarks as rejections of your ideas as steps towards improved rendering.  I am just trying to point out potential drawbacks.

Treating entrances, address nodes and addresses on buildings exactly the same would have the advantage of simplicity.  It could, however - in absence of de-duplication - have the effect of encouraging mappers to selectively tag some address components on entrances and not others - which could make data maintainance and interpretation more difficult.

De-duplication - including transferring data from certain tags from entrances and address nodes to the building they are on - could solve these things.  But i am not sure if this is advisable (in terms of being intuitive for mappers) and viable (in terms of performance).

Bottom line: Your proposal might be ok considering the alternatives are not better.

**Comment by dch0ph** [comment]
I've extended the mockup:

<img width="424" height="186" alt="Image" src="https://github.com/user-attachments/assets/48718e51-79e6-47d4-9cc6-2224a151c54b" />

to clarify that I'm only suggesting dropping `addr:housename` on entrance labels containing "subunit" tagging. 

> > Entrances are only labelled from Z18.
> 
> That would mean discouraging tagging addresses on entrances vs. tagging on buildings/address nodes.

Yes. That would need further input. Personally I think it's logical to postpone labelling of entrances given that the label will be harder to place and read compared to the middle of a building. 

> Tagging addresses only on entrance nodes is widely considered perfectly valid mapping

It's hard to distinguish between "mapping address only on entrance node" and "duplicating address information so that entrance node is a complete postal address". I would guess, for example, that `ref` on entrance is almost exclusively the second type. 

> I see that we are a bit in Yak shaving mode here, coming from shortening lengthy number lists to differentiating different address components

Indeed. It would be useful to separate things things we can take forward to #5272 from a wider discussion of address labelling, which would benefit from further input.

I think a quick win, as implied by the mockup, would be to address #3038 by using a small `text-dy` to offset the address label when an entrance is present.

**Comment by imagico** [comment]
> I've extended the mockup:
> 
> <img alt="Image" width="424" height="186" src="https://private-user-images.githubusercontent.com/20069699/627968652-48718e51-79e6-47d4-9cc6-2224a151c54b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODUzMTczODEsIm5iZiI6MTc4NTMxNzA4MSwicGF0aCI6Ii8yMDA2OTY5OS82Mjc5Njg2NTItNDg3MThlNTEtNzllNi00N2Q0LTljYzYtMjIyNGExNTFjNTRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzI5VDA5MjQ0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiODNmNGZlM2Y2MmQyYTMzNDhkOWZiN2MzZjkwNjAwN2ZkZGVjNmQxODcwMmJkMTRlNzk2ZjFiY2RiZjBiMjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.xicR8zJTtlGnqwS5Jsd-9KWo0toR_fJP5jNOrM085cs">
> to clarify that I'm only suggesting dropping `addr:housename` on entrance labels containing "subunit" tagging.

How do you intend to label a combination of `addr:housenumber`, `addr:housename` and `addr:flats` on a building or address node?

A possible alternative would be to label entrances with `addr:housenumber` and `addr:housename` with *housenumber (housename)*.  That would use added punctuation with all the drawbacks, but the combination is rare in tagging and it would communicate the housename as something clearly secondary on the entrance.

> Yes. That would need further input. Personally I think it's logical to postpone labelling of entrances given that the label will be harder to place and read compared to the middle of a building.

I think one of the reasons why some mappers prefer addresses on entrance nodes is that it often avoids the collision between POIs in the building and the address labels.  (At the same time one of the reasons to avoid tagging entrance nodes with addresses is #3038)

> > Tagging addresses only on entrance nodes is widely considered perfectly valid mapping
> 
> It's hard to distinguish between "mapping address only on entrance node" and "duplicating address information so that entrance node is a complete postal address". I would guess, for example, that `ref` on entrance is almost exclusively the second type.

This would be something that could be determined through data analysis though - in case someone finds the time to do this.  The questions of interest are:

* what percentage of entrance nodes with address tags are on buildings which have address tags themselves of contain an address node with address tags?
* what percentage of these have certain address components only on the entrance and not on the other features?
* what percentage of these have all the address components from the building/address node also on the entrance?

**Comment by dch0ph** [comment]
> How do you intend to label a combination of `addr:housenumber`, `addr:housename` and `addr:flats` on a building or address node?

<img width="67" height="34" alt="Image" src="https://github.com/user-attachments/assets/2f309841-32f2-43ce-a5c2-08273cf101b6" />

But this is a 3-part label, so lots to go wrong there. I think the "mapnik anchors", so that rendering is conditional on successful placement/rendering of an earlier item, is the single most useful feature that could be added to mapnik.

I am becoming doubtful that "simpler rendering for entrances" would be a cheap-and-cheerful way to address duplication given the prevalence of "addresses on entrances only" tagging. 

But I do like the "systematic displacement" option to avoid the ugliness of `[]` and the legibility issues of "proper" sub/superscripting.
